src/app/data-model-provider/MetadataList.cpp

```diff
@@ -25,106 +25,157 @@ namespace app {
+    mElementCount      = other.mElementCount;
+    mCapacity          = other.mCapacity;
+    mBufferIsAllocated = other.mBufferIsAllocated;
+
+    // clear other
+    other.mBuffer            = nullptr;
+    other.mElementCount      = 0;
+    other.mCapacity          = 0;
+    other.mBufferIsAllocated = false;
 
-    if ((mBuffer != nullptr) && mAllocated)
+    return *this;
+}
+
+CHIP_ERROR GenericAppendOnlyBuffer::EnsureAppendCapacity(size_t numElements)
+{
+    if (mCapacity >= mElementCount + numElements)
     {
-        chip::Platform::MemoryFree(mBuffer);
+        // Sufficient capacity already exists
+        return CHIP_NO_ERROR;
     }
 
-    mBuffer = static_cast<uint8_t *>(chip::Platform::MemoryCalloc(numElements, mElementSize));
+    if (mBuffer == nullptr)
+    {
+        mBuffer            = static_cast<uint8_t *>(Platform::MemoryCalloc(numElements, mElementSize));
+        mCapacity          = numElements;
+        mBufferIsAllocated = true;
+        return mBuffer != nullptr ? CHIP_NO_ERROR : CHIP_ERROR_NO_MEMORY;
+    }
 
-    VerifyOrReturnError(mBuffer != nullptr, CHIP_ERROR_NO_MEMORY);
+    // we already have the data in buffer. we have two choices:
+    //   - allocated buffer needs to be extended
+    //   - re-used const buffer needs to be copied over
+    if (mBufferIsAllocated)
+    {
+        auto new_buffer = static_cast<uint8_t *>(Platform::MemoryRealloc(mBuffer, (mElementCount + numElements) * mElementSize));
+        VerifyOrReturnError(new_buffer != nullptr, CHIP_ERROR_NO_MEMORY);
+        mBuffer = new_buffer;
+    }
+    else
+    {
+        // this is NOT an allocated buffer, but it should become one
+        auto new_buffer    = static_cast<uint8_t *>(Platform::MemoryCalloc(mElementCount + numElements, mElementSize));
+        mBufferIsAllocated = true;
+        VerifyOrReturnError(new_buffer != nullptr, CHIP_ERROR_NO_MEMORY);
+        memcpy(new_buffer, mBuffer, mElementCount * mElementSize);
+        mBuffer = new_buffer;
+    }
+    mCapacity = mElementCount + numElements;
 
-    mAllocated = true;
-    mCapacity  = numElements;
     return CHIP_NO_ERROR;
 }
 
-CHIP_ERROR GenericMetadataList::AppendRaw(const void * buffer)
+CHIP_ERROR GenericAppendOnlyBuffer::AppendSingleElementRaw(const void * buffer)
 {
-    VerifyOrReturnError(!mIsImmutable, CHIP_ERROR_INCORRECT_STATE);
-    VerifyOrReturnError(mElementCount < mCapacity, CHIP_ERROR_NO_MEMORY);
+    VerifyOrReturnError(mElementCount < mCapacity, CHIP_ERROR_BUFFER_TOO_SMALL);
     memcpy(mBuffer + mElementCount * mElementSize, buffer, mElementSize);
     mElementCount++;
     return CHIP_NO_ERROR;
 }
 
-void GenericMetadataList::Invalidate()
+CHIP_ERROR GenericAppendOnlyBuffer::AppendElementArrayRaw(const void * buffer, size_t numElements)
 {
-    if ((mBuffer != nullptr) && mAllocated)
+    ReturnErrorOnFailure(EnsureAppendCapacity(numElements));
+
+    memcpy(mBuffer + mElementCount * mElementSize, buffer, numElements * mElementSize);
```

Comment:
Should be memmove, because nothing guarantees that `buffer` does not point somewhere inside `mBuffer` here.

Reply:
That would be bad, because EnsureAppendCapacity may reallocate/move data, so if the buffer we receive points inside itself, it would not be moved.

In theory possible (e.g. by calling EnsureAppendCapacity first) however it seems like a pattern we should not support. I would rather place this in documentation/requirements because of that. It is a footgun to try to append from itself. Also some logical error because we mostly want unique things in our array.

Reply:
Probaly in the current design not even possible because the builders do not provide external access to the buffer. Adding docs and `__restrict__`

Reply:
OK.